torch/_higher_order_ops/scan.py

```diff
@@ -0,0 +1,339 @@
+# mypy: allow-untyped-defs
+import functools
+import itertools
+from typing import Callable, List
+
+import torch
+import torch._prims_common as utils
+import torch._subclasses.functional_tensor
+import torch.utils._pytree as pytree
+from torch._C import DispatchKey
+from torch._higher_order_ops.utils import (
+    _set_compilation_env,
+    autograd_not_implemented,
+    reenter_make_fx,
+    unique_graph_id,
+)
+from torch._ops import HigherOrderOperator
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.fx.experimental.proxy_tensor import (
+    disable_proxy_modes_tracing,
+    ProxyTorchDispatchMode,
+    track_tensor_tree,
+)
+from torch.utils._python_dispatch import _get_current_dispatch_mode
+
+
+aten = torch._ops.ops.aten
+
+
+def wrap_combine_fn_flat(*args, combine_fn, spec, num_leaves):
+    assert len(args) == 2 * num_leaves
+    lhs = pytree.tree_unflatten(args[:num_leaves], spec)
+    rhs = pytree.tree_unflatten(args[num_leaves:], spec)
+    combined = combine_fn(lhs, rhs)
+    combined_leaves = pytree.tree_leaves(combined)
+    assert num_leaves == len(combined_leaves)
+    return combined_leaves
+
+
+def scan(
```

Comment:
In fact, we might want to support the case where combined_output is different from carry: i.e.
```python
# This is our current semantic
def combine_fn(init, input):
  # next iteration's init is current iteration's output
   return init + input
```

However, there are cases that the carry is different from output e.g.
```
def combind_fn(init, input):
  # next iteration's input is init + input
  # but current iteration's output is init
  return init + input, init
```
Another jax example would be:
```
import jax.numpy as jnp
from jax import lax

def body(init, x):
    return (init + x, init + x.sum())

xs = jnp.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]])
init = jnp.array([0, 0, 0, 0], dtype=xs.dtype)
result = lax.scan(body, init, xs)
print(result)
```

This is kind of bigger change that we need to do.

Reply:
The other implication of this change is that we probably want to split the tests of associative with scan. In the future, we probably want to extend associative_scan to have the same interface of scan instead of the other way around.

Reply:
I added a [TODO](https://github.com/pytorch/pytorch/blob/810491fddfb4e51cdbe4e56c7094c8a56729c5bd/torch/_higher_order_ops/scan.py#L98) for this new user interface

Reply:
I think we need to do it quickly (best in this pr). I'm also OK with do it in the next pr since the current PR is too large. Autograd PR would require and we really don't want to expose the current interface to users.